### ISSUE TITLE
Preprocess using both cppo and PPXs

### ISSUE THREAD

**Issue by aantron** [janestreet/jbuilder#171]
From https://github.com/ocsigen/lwt/issues/411#issuecomment-311214821:

> it is possible to both apply a PPX and preprocess with cppo, but it requires the [separate] cppo `rule` stanza, and so presumably a source file with a name like `foo.cppo.ml`. It would be nice if one could both apply cppo and PPXs to source named `foo.ml`.

> This is not a high priority, but it would make files like `lwt_unix.cppo.ml`, where we both want to use cppo, and apply Bisect_ppx, ever so slightly more maintainable.

I'm basically thinking of being able to have either multiple `preprocess` stanzas, or list both `pps` and [`action`](http://jbuilder.readthedocs.io/en/latest/quick-start.html#using-cppo) in one `preprocess` stanza. There may, of course, be other, better ways.

**Comment by aantron** [comment]
Looking at ocsigen/lwt#411 closer, the issue with paths appears to be an inconsistency between how Jbuilder runs PPXs and how it [suggests](https://github.com/janestreet/jbuilder/blob/3b1182c23a0c1993e4077a2cb5e3a611697806e5/doc/quick-start.rst#using-the-cppoml-style-like-the-ocamlbuild-plugin) to run cppo. I believe this is what @andrewray [found](https://github.com/ocsigen/lwt/issues/411#issuecomment-311935687).

```scheme
(rule
 ((targets (lwt_unix.ml))
  (deps    (lwt_unix.cppo.ml))
  (action
   (run ${bin:cppo} -V OCAML:${ocaml_version} ${<} -o ${@}))))
```

Presumably, the above causes Jbuilder to pass the filename `lwt_unix.cppo.ml` to cppo, so of course cppo has to use that. This filename is relative to the `src/unix/` subdirectory.

When the PPX is later run, the working directory is (a copy of) the root directory of the project, so the path to access this file is `src/unix/lwt_unix.cppo.ml` instead.

It would be nice to eliminate this inconsistency (as well as allow running cppo and PPXs from the `preprocess` stanzas).

Lwt is working around this like so:

```scheme
(rule
 ((targets (lwt_unix.ml))
  (deps    (lwt_unix.cppo.ml))
  (action
   (chdir ${ROOT} (run ${bin:cppo} -V OCAML:${ocaml_version} ${<} -o ${@})))))
```

as suggested by @andrewray.

**Comment by diml** [comment]
Indeed. However this is a complex problem and I'm not sure what's the best way to solve it. It seems clear to me that if you write a command in `foo/bar/jbuild`, then it should be run from `foo/bar`. It seems more intuitive, especially given that with jbuilder the relative path from the compilation root is not constant.

By extension, we should run all commands derived from stanzas in `foo/bar/jbuild`  in `foo/bar`. However, if we did that the locations reported by the compiler in error messages would contain only the base names and editors might not be able to properly jump to the source location. We could fix this by implementing the suggestion in #138. However, this wouldn't help for locations at runtime (assertion failures, backtraces, debugging in gdb, ...). Though the problem is even more complicated for runtime locations.

As a workaround, jbuilder runs various commands generated internally from the root of the workspace and the `(chdir ${ROOT} ...)` trick is necessary. It is mentioned in the manual and should be used in the quickstart as well.

**Comment by diml** [comment]
BTW, for the initial problem, if you use `ppx_driver`, you can pass a `-pp <cmd>`  parameter. A similar parameter could be added to the omp driver so that one could write:

```scheme
(preprocess (pps (ppx_foo -pp=${bin:cppo})))
```

This would be quicker to implement than this PR.

**Comment by aantron** [comment]
@let-def Is this

> A similar parameter could be added to the omp driver so that one could write:
> ```
> (preprocess (pps (ppx_foo -pp=${bin:cppo})))
> ```

something that should be added to omp?

**Comment by diml** [comment]
BTW, have you looked at [ppx_optcomp](https://github.com/janestreet/ppx_optcomp)? It provides similar features to cppo, but as a pure ppx. Text preprocessors tends to not work with the rest of the tooling (merlin, ocamlformat, ...), so I'm not sure that's the right direction to push

**Comment by aantron** [comment]
ppx_optcomp works in Lwt. ppx_optcomp depends on Base though, and I don't think Lwt is ready to depend on Base transitively for a few lines of conditional compilation.

I agree that cppo is perhaps to not worth supporting in Dune. We'll eventually find another way to do the preprocessing we need for Lwt.

**Comment by diml** [comment]
One way that might work is to extract the few elements that needs cppo into separate files and have a rule to copy the right file in place. We do that in Base to support multiple versions of OCaml.

**Comment by Leandros** [comment]
Any progress on this?

**Comment by Leandros** [comment]
Any progress on this?

**Comment by diml** [comment]
No. However, I still not convinced by the motivations here. We can avoid foreign syntaxes entirely with extension points, so using extension/attribute based conditional compilation just seems better. IIUC, the only reason one would prefer cppo to ppx_optcomp is becasue cppo has less dependencies, but at the same time if you want to use any ppx rewriter, then they all have much more dependencies that cppo.

**Comment by aantron** [comment]
Besides the dependency issue, there is also that PPXs depend on ocaml-migrate-parsetree (and maybe ppx_tools_versioned), which are sometimes slow to upgrade to new OCaml versions.

For example, in https://github.com/ocaml/odoc/issues/316, we loosened how odoc depends on Bisect_ppx, in order to avoid being held back by OMP. odoc also uses cppo. If we were instead using ppx_optcomp in addition to Bisect_ppx, we would remain held back for now. In Lwt, only `lwt_ppx` is held back by OMP. But, if we used ppx_optcomp, all of Lwt would be held back.

I personally don't think this is solved by using cppo, but by helping OMP to get updated faster. However, it is another factor to consider, that cppo requires less maintenance when new OCaml versions are released.

Regarding dependencies, odoc is another project that uses cppo, but probably shouldn't choose to transitively depend on `base`, `stdio`, etc. We recently removed much smaller dependencies: https://github.com/ocaml/odoc/pull/305.

I'll take a look at the code of ppx_optcomp, and see if it is reasonable to remove some of its deps.

**Comment by diml** [comment]
For the long term, I have a few ideas to make the ppx ecosystem completely pain free. I.e. really have a stable API for ppxs to rely on so that newer compilers never break existing ppx rewriters. But that will take some time.

In the meantime, I'm still not keen on investing time in complicating the preprocessing pipeline of dune. Especially given that there is a workaround: you can define manual rules to call cppo on the few files you want to preprocess.

**Comment by domq** [comment]
Another use case is when I have toplevel directives in my source code. I'd like to be able to both C-c C-b my file, use `inline_tests`, and compile it with dune; unfortunately the following doesn't work.

```
(preprocess (action (run sed "s/^#.*;;//")) (pps ppx_inline_test))
```

**Comment by CXWorks** [comment]
Feature request +1, any ways to bypass it? Thx
